scipy/spatial/ckdtree.pyx

```diff
@@ -887,32 +884,24 @@ cdef class cKDTree:
         """
 
         cdef:
-            const np.float64_t[::1] vrr
-            const np.float64_t[:, ::1] vxx
             object[::1] vout
             np.intp_t[::1] vlen
-            list tmp
-            np.intp_t i, j, n, m
-            np.intp_t xndim
-
-        xshape = np.shape(x)
-        if xshape[-1] != self.m:
-            raise ValueError("Searching for a %d-dimensional point in a "
-                             "%d-dimensional KDTree" %
-                                 (int(xshape[-1]), int(self.m)))
-
-        vxx = np.ascontiguousarray(x, dtype=np.float64).reshape(-1, self.m)
-        vrr = np.ascontiguousarray(np.broadcast_to(r, xshape[:-1]), dtype=np.float64).reshape(-1)
-
-        retshape = xshape[:-1]
+            np.ndarray x_arr = np.ascontiguousarray(x, dtype=np.float64)
+            ckdtree *cself = self.cself
+            bool rlen = return_length
+            # compatibility with the old bug not sorting scalar queries.
+            bool sort_output = return_sorted or (
+                return_sorted is None and x_arr.ndim > 1)
 
-        # scalar query if xndim == 1
-        xndim = len(xshape)
+            np.intp_t n = num_points(x_arr, cself.m)
+            tuple retshape = np.shape(x_arr)[:-1]
+            np.ndarray r_arr = broadcast_contiguous(r, shape=retshape,
+                                                    dtype=np.float64)
 
-        # allocate an array of std::vector<npy_intp>
-        n = np.prod(retshape)
```

Comment:
I'm seeing a number of errors related to this line (and a few others) in this old version of the Cython source file when using NumPy pre-release wheels in gh-12906.

Here's an example failure log: https://travis-ci.org/github/scipy/scipy/jobs/733853103

Any thoughts?

Reply:
Related to certain versions of numpy treats a tuple (,)?

retshape = xshape[:-1] and xshape is a python tuple, so I am not sure how it could be plausibly understood as a float scalar without something lower level has gone wrong.

Reply:
Maybe numpy.shape is no longer returning a tuple?

Reply:
I've been trying to look into this but can't make any sense of it:
- In gdb, `prod`'s return has `ob_type` `PyDoubleArrType_Type` which sounds like float64 which is consistent with the error, however it's the same in versions that don't fail. Maybe this is just a quirk of numpy scalars?
- Tests pass on everything except when pre-release numpy _and_ pre-release cython are combined, in which case it errors.
- `np.prod(np.shape(array))` from python always returns a scalar `int64` value.

I had hoped that I could at least minimise it to something like
```python
def test_func(a):
    cdef np.intp_t n = np.prod(np.shape(a))
    return n
```

But for some reason that doesn't fail. I strongly suspect this is a cython bug, but can't narrow it down to be sure.

Reply:
Did you check the value of retshape in the failed case? It shouldn't be a scalar, but I wonder if it is empty? If that's the case np.prod should have returned 1, but may have deduced the type as float64. A stricter version of Cython or numpy may start to complain about the casting from float64 to int.

If that's the case, a few lines above, change it to retshape = np.asarray(xshape[1:], dtype='intp') can solve the problem. 

There is likely other occurrences in the file that needs to be fixed too.

Reply:
Yes! That's exactly it. NumPy actually always deduced float from an empty tuple and Cython has only now become more strict and started rejecting floats.

Rather than doing `asarray`, I suggest `np.prod(xshape, dtype=np.intp)`.

Reply:
Thanks both, this was most helpful in getting it patched locally with bleeding-edge Cython.